File: 0146-lru-cache/0146-lru-cache.py
class CacheItem:
    def __init__(
        self,
        key: int,
        val: int,
        prevPtr: "CacheItem" = None,
        nextPtr: "CacheItem" = None,
    ) -> None:
        self.key = key
        self.val = val
        self.nextPtr = nextPtr
        self.prevPtr = prevPtr

    def __str__(self) -> str:
        return f"[CacheItem(prev={self.prevPtr}, key={self.key}, val={self.val}, next={self.nextPtr.key}]"


class CacheQueue:
    def __init__(self) -> None:
        self.head = None
        self.tail = None
        self.capacity = 0

    def appendleft(self, key: int, val: int) -> CacheItem:
        item = CacheItem(key, val)
        # increment size
        self.capacity += 1
        # if head is not present
        if not self.head:
            self.head = item
            self.tail = item
            return item
        # if head is present
        item.nextPtr = self.head
        self.head.prevPtr = item
        self.head = item
        return item

    def remove(self, item: CacheItem) -> tuple[int, int]:
        # trying to remove when queue is empty
        if self.capacity == 0:
            raise RuntimeError(f"Invalid operation remove for item: {item}")

        nextItem = item.nextPtr
        prevItem = item.prevPtr
        # if prev item is present connect to item.next
        if prevItem:
            prevItem.nextPtr = nextItem
        # else its head so update head
        else:
            self.head = nextItem
        # if next item is present
        if nextItem:
            nextItem.prevPtr = prevItem
        else:
            self.tail = prevItem
        # make sure head prev is None
        if self.head:
            self.head.prevPtr = None
        # make sure tail next is None
        if self.tail:
            self.tail.nextPtr = None
        # cleanup
        item.nextPtr = None
        item.prevPtr = None
        # decrement size
        self.capacity -= 1
        return item.key, item.val

    def pop(self) -> CacheItem:
        # if queue is empty return none
        if self.capacity == 0:
            return None

        prevTail = self.tail
        self.tail = prevTail.prevPtr
        if self.tail:
            self.tail.nextPtr = None
        else:
            # only item removed
            self.head = self.tail = None
        # cleanup
        prevTail.prevPtr = None
        # decrement capacity
        self.capacity -= 1
        return prevTail

    def movetohead(self, item: CacheItem) -> CacheItem:
        # removing item
        key, val = self.remove(item)
        # moving to head (head has higher priority)
        return self.appendleft(key, val)

    def size(self) -> int:
        return self.capacity


class LRUCache:
    def __init__(self, capacity: int) -> None:
        self.capacity = capacity
        self.cacheq = CacheQueue()
        self.mem: dict[int, CacheItem] = {}

    def get(self, key: int) -> int:
        if key in self.mem:
            item = self.mem[key]
            # recently accessed
            self.mem[key] = self.cacheq.movetohead(item)
            return self.mem[key].val
        # else item is not present
        return -1

    def put(self, key: int, val: int) -> None:
        # if new item
        if key not in self.mem:
            # if cacheq size limit reached then pop from tail
            # to create space
            if self.cacheq.size() == self.capacity:
                poppedItem = self.cacheq.pop()
                # delete from map as well
                del self.mem[poppedItem.key]
            # insert item at head
            item = self.cacheq.appendleft(key, val)
            self.mem[item.key] = item
        else:
            item = self.mem[key]
            # update value
            item.val = val
            # recently accessed
            self.mem[key] = self.cacheq.movetohead(item)

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.java
class LRUCache {
    class CacheNode {
        public CacheNode prev;
        public int key;
        public int value;
        public CacheNode next;

        public CacheNode(CacheNode prev, int key, int value, CacheNode next) {
            this(key, value);
            this.prev = prev;
            this.next = next;
        }

        public CacheNode(int key, int value) {
            this.key = key;
            this.value = value;
            this.prev = null;
            this.next = null;
        }
    }

    class CacheQueue {
        public CacheNode head;
        public CacheNode tail;
        public int size;

        public CacheQueue() {
            this.head = null;
            this.tail = null;
            this.size = 0;
        }

        public CacheNode appendLeft(int key, int val) {
            CacheNode newNode = new CacheNode(key, val);
            if (isEmpty()) {
                this.size += 1;
                this.head = newNode;
                this.tail = newNode;
                return newNode;
            }
            this.size += 1;
            newNode.next = this.head;
            this.head.prev = newNode;
            this.head = newNode;
            return newNode;
        }

        public CacheNode remove(CacheNode node) {
            if (isEmpty() || node == null) {
                return null;
            }
            CacheNode prev = node.prev;
            CacheNode next = node.next;
            if (prev != null) {
                prev.next = next;
            } else {
                this.head = next;
            }
            if (next != null) {
                next.prev = prev;
            } else {
                this.tail = prev;
            }
            if (this.head != null) {
                this.head.prev = null;
            }
            if (this.tail != null) {
                this.tail.next = null;
            }
            this.size -= 1;
            node.prev = null;
            node.next = null;
            return node;
        }

        public CacheNode pop() {
            if (isEmpty()) {
                return null;
            }
            CacheNode tailNode = remove(tail);
            return tailNode;
        }

        public CacheNode moveToHead(CacheNode node) {
            if (isEmpty()) {
                return null;
            }
            CacheNode removedNode = remove(node);
            return appendLeft(removedNode.key, removedNode.value);
        }

        public int length() {
            return this.size;
        }

        public boolean isEmpty() {
            return this.size == 0;
        }
    }

    private Map<Integer, CacheNode> mem;
    private CacheQueue cacheQueue;
    private int capacity;

    public LRUCache(int capacity) {
        this.mem = new HashMap<>();
        this.cacheQueue = new CacheQueue();
        this.capacity = capacity;
    }

    public int get(int key) {
        if (!mem.containsKey(key)) {
            return -1;
        }
        CacheNode node = mem.get(key);
        mem.put(key, cacheQueue.moveToHead(node));
        return mem.get(key).value;
    }

    public void put(int key, int value) {
        if (!mem.containsKey(key)) {
            if (cacheQueue.length() == capacity) {
                CacheNode popped = cacheQueue.pop();
                mem.remove(popped.key);
            }
            mem.put(key, cacheQueue.appendLeft(key, value));
        } else {
            CacheNode node = mem.get(key);
            node.value = value;
            mem.put(key, cacheQueue.moveToHead(node));
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */